ksql-plugins/ksql-rocksdb-config-setter/src/main/java/io/confluent/ksql/rocksdb/KsqlBoundedMemoryRocksDBConfig.java

```diff
@@ -0,0 +1,132 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rocksdb;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.kafka.streams.state.RocksDBConfigSetter;
+import org.rocksdb.BlockBasedTableConfig;
+import org.rocksdb.Options;
+
+/**
+ * This {@code RocksDBConfigSetter} implementation limits the total memory used
+ * across all RocksDB instances to the number of bytes passed via
+ * "ksql.plugins.rocksdb.total.memory", and also configures
+ * the shared RocksDB thread pool to use "ksql.plugins.rocksdb.num.background.threads" threads.
+ * </p>
+ * See https://docs.confluent.io/5.3.0/streams/developer-guide/memory-mgmt.html#rocksdb.
+ */
+public class KsqlBoundedMemoryRocksDBConfig implements RocksDBConfigSetter {
+
+  private static final String CONFIG_PREFIX = "ksql.plugins.rocksdb.";
+  private static final String TOTAL_OFF_HEAP_MEMORY_CONFIG = CONFIG_PREFIX + "total.memory";
+  private static final String N_BACKGROUND_THREADS_CONFIG =
+      CONFIG_PREFIX + "num.background.threads";
+
+  private static final double INDEX_FILTER_BLOCK_RATIO = 0.1;
+  private static final int N_MEMTABLES = 6;
+  private static final long BLOCK_SIZE = 4096L;
+
+  private static org.rocksdb.Cache cache;
+  private static org.rocksdb.WriteBufferManager writeBufferManager;
+
+  private static AtomicBoolean configured = new AtomicBoolean(false);
+  private static long totalOffHeapMemory;
+  private static long memtableSize;
+  private static long totalMemtableMemory;
+  private static int numBackgroundThreads;
+
+  public static void configure(final Map<String, Object> config) {
+    configure(config, new Options());
+  }
+
+  @VisibleForTesting
+  static void configure(final Map<String, Object> config, final Options options) {
+    if (configured.getAndSet(true)) {
+      throw new IllegalStateException(
+          "KsqlBoundedMemoryRocksDBConfig has already been configured. Cannot re-configure.");
+    }
+
+    try {
+      limitTotalMemory(config);
+      configureNumThreads(config, options);
+    } catch (IllegalArgumentException e) {
+      configured.set(false);
+      throw e;
+    }
+  }
+
+  @VisibleForTesting
+  static void reset() {
+    configured.set(false);
+  }
+
+  private static void limitTotalMemory(final Map<String, Object> config) {
+    if (!config.containsKey(TOTAL_OFF_HEAP_MEMORY_CONFIG)) {
+      throw new IllegalArgumentException(
+          "Missing configuration: " + TOTAL_OFF_HEAP_MEMORY_CONFIG);
+    }
+    totalOffHeapMemory = Long.parseLong((String)config.get(TOTAL_OFF_HEAP_MEMORY_CONFIG));
+
+    memtableSize = totalOffHeapMemory / 2 / N_MEMTABLES;
+    totalMemtableMemory = N_MEMTABLES * memtableSize;
```

Comment:
Where is this math coming from? This is going to create massive memtables on configs with more memory. Let's keep the memtable size the same as streams (16MB), and just set a cap on total memtable memory. What you have here (totalOffHeapMemory / 2) seems fine. Whatever isn't used should just get used for the read cache.

Reply:
Is this because we don't know how many write buffers we'll wind up needing? (since each rocks needs at least 2)?

Reply:
I made the change to reconcile the fact that we have a variable total write buffer memory (passed as the first argument when creating the `WriteBufferManager`), yet Streams by default sets write buffer size to 16MB and max write buffer number to 3. Unless I've misunderstood, without either increasing write buffer size or write buffer number from the Streams default, we'll never use more than 48MB for the write buffers, despite our desire for that limit to be configurable.

Reply:
Just want to chime in quickly here, the memtable size (and max number) is still per rocks instance. So each instance could in theory use up to 48MB, but if we limit the total write buffer memory the WriteBufferManager will flush some of these to make sure the total across instances stays within the bounds you set.
Also, write buffer size is actually a minimum size to be a candidate for flushing, not a cap. But in most cases it would likely end up being around this size before being made immutable and/or flushing.